### DISCUSSION TITLE
RPM build setup

### DISCUSSION THREAD

**Discussion by Zomtir** [coin-or/Ipopt#475] [General]
I'm interesting in packaging up IPOPT as rpm to distribute it across different machines at a higher convenience level compared to using "make install" and distributing the build folder.

For CentOS7 there is an RPM build available via EPEL [1]. Unfortunately no Coin-or package has found its way into the CentOS8 EPEL or powertools repository yet.

Would it be possible to share the workflow how IPOPT was packaged into EPEL for CentOS7? I'd be especially glad about the spec file and the configure flags.

I managed to make the program on the latest CentOS8 Stream with the following flags (after having installed the necessary dependencies from EPEL/powertools [2]):

```
mkdir build && cd build
../configure --with-lapack --with-mumps --with-mumps-cflags="-I/usr/include/MUMPS" --with-mumps-lflags="-lmumps_common -lcmumps -ldmumps" --without-hsl --without-asl
make
```

[1] https://centos.pkgs.org/7/epel-aarch64/coin-or-Ipopt-3.12.8-5.el7.aarch64.rpm.html
[2] blas-devel lapack-devel metis-devel MUMPS-devel

**Comment by svigerske** [comment]
The Changelog in [1] seems to be mention the e-mail address of someone who seems to have maintained this package. Maybe they have the files you mention.

From your configure flags, only
```
--with-mumps-cflags="-I/usr/include/MUMPS" --with-mumps-lflags="-lmumps_common -ldmumps"
```
should be necessary.

Ipopt itself does not interface to Metis, so you probably don't need metis-devel.

Run also make test to check whether Ipopt works with this build of MUMPS.

**Reply by Zomtir** [reply]
Thanks for the tips! The suggested configure flags and unit tests worked flawlessly (appended). I'll write a mail to Antonio Trande and ask if he's willing to help out :)

```
[root@workstation build]# make test
Making all in src
make[1]: Entering directory '/root/Ipopt/build/src'
make  all-recursive
make[2]: Entering directory '/root/Ipopt/build/src'
make[3]: Entering directory '/root/Ipopt/build/src'
make[3]: Nothing to be done for 'all-am'.
make[3]: Leaving directory '/root/Ipopt/build/src'
make[2]: Leaving directory '/root/Ipopt/build/src'
make[1]: Leaving directory '/root/Ipopt/build/src'
Making all in contrib/sIPOPT
make[1]: Entering directory '/root/Ipopt/build/contrib/sIPOPT'
Making all in src
make[2]: Entering directory '/root/Ipopt/build/contrib/sIPOPT/src'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/root/Ipopt/build/contrib/sIPOPT/src'
make[2]: Entering directory '/root/Ipopt/build/contrib/sIPOPT'
make[2]: Nothing to be done for 'all-am'.
make[2]: Leaving directory '/root/Ipopt/build/contrib/sIPOPT'
make[1]: Leaving directory '/root/Ipopt/build/contrib/sIPOPT'
make[1]: Entering directory '/root/Ipopt/build'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/root/Ipopt/build'
cd test; make test
make[1]: Entering directory '/root/Ipopt/build/test'
  CXX      hs071_main.o
  CXX      hs071_nlp.o
  CXXLD    hs071_cpp
  CC       hs071_c.o
  CCLD     hs071_c
  CXX      emptynlp.o
  CXXLD    emptynlp
  CXX      getcurr.o
  CXXLD    getcurr
ln -s ../examples/hs071_f/hs071_f.f hs071_f.f
  F77      hs071_f.o
  F77LD    hs071_f
  CXX      parametricTNLP.o
  CXX      parametric_driver.o
  CXXLD    parametric_cpp
  CXX      MySensTNLP.o
  CXX      redhess_cpp.o
  CXXLD    redhess_cpp
javac -classpath ../src/org.coinor.ipopt.jar ../../examples/hs071_java/HS071.java -d /root/Ipopt/build/test
chmod u+x ./run_unitTests
./run_unitTests
 
Running unitTests...
 
Testing AMPL Solver Executable...
    no AMPL solver executable found, skipping test...
Testing C++ Example...
    Test passed!
Testing C Example...
    Test passed!
Testing Fortran Example...
    Test passed!
Testing Java Example...
    Test passed!
Testing sIpopt Example parametric_cpp...
    Test passed!
Testing sIpopt Example redhess_cpp...
    Test passed!
Testing EmptyNLP Example...
    Test passed!
Testing GetCurr Example...
    Test passed!
make[1]: Leaving directory '/root/Ipopt/build/test'
```

**Comment by sagitter** [comment]
Hello.

@Zomtir,

you can see [here](https://src.fedoraproject.org/rpms/coin-or-Ipopt/blob/rawhide/f/coin-or-Ipopt.spec) the spec file used for building Ipopt.
Unfortunately, [mp](https://src.fedoraproject.org/rpms/mp) dependency is missing in EPEL8, even if its source rpm is ready to be used in a EPEL8 buildroot.

**Reply by Zomtir** [reply]
@sagitter This was great insight and I'm honestly impressed by the cascading koji build chain. I tried using the rawhide spec file without ASL support (%global with_asl 0), because this was the only feature depending on mp.

After installing the remaining dependencies, it failed at creating the help manual with help2man:

```
+ help2man -N /root/rpmbuild/BUILDROOT/coin-or-Ipopt-3.13.0-11.el8.x86_64/usr/bin/ipopt
help2man: can't get `--help' info from /root/rpmbuild/BUILDROOT/coin-or-Ipopt-3.13.0-11.el8.x86_64/usr/bin/ipopt
Try `--no-discard-stderr' if option outputs to stderr
error: Bad exit status from /var/tmp/rpm-tmp.Umsc9Z (%install)


RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.Umsc9Z (%install)
```

I'll now spend some more time understanding where things are going wrong. Nonetheless it was unexpectedly easy and straightforward to get the rpmbuild rolling up to this point.

I noticed you pushed some spec update to mp for Fedora 35 yesterday, does this mean IPOPT support is eventually on the EPEL8 horizon anyway?

**Comment by sagitter** [comment]
>I'll now spend some more time understanding where things are going wrong. Nonetheless it was unexpectedly easy and straightforward to get the rpmbuild rolling up to this point.

Probably, something is missed in your buildroot. Use `# dnf builddep coin-or-Ipopt.spec` to be sure that all needed packages are installed before compiling.
`coin-or-Ipopt` package is correctly built in EPEL8 (including `mp`): https://copr.fedorainfracloud.org/coprs/sagitter/ForTesting/builds/

> I noticed you pushed some spec update to mp for Fedora 35 yesterday, does this mean IPOPT support is eventually on the EPEL8 horizon anyway?

Please, open a ticket in [RedHat Bugzilla](https://bugzilla.redhat.com/) to require Ipopt in EPEL8.

**Comment by Zomtir** [comment]
I managed to put together a working spec file. Do not use this, it is horrible for various reasons:

- The documentation was thrown out and the the license file was put in the main package. The manpages **are** depending on  ASL because it copies the manual from the ASL build. The manpage itself is a 5-liner and does not really help anyway. Doxygen is disabled by upstream by default.
- MPI support was dropped. It required more patching than I felt comfortable with and there were some internal library linking issues I could not solve.
- ASL support was dropped. The mp package dependency also required patches and did not have a regular release since 2016. After manually cloning the repository, creating a tarball without the git files and fixing the commit hashes, I still failed at applying the patches as part of the spec file.
- Install paths are hardcoded. By default Ipopt is installed into /usr/local/[lib,include], but rpmbuild build macros heavily imply /usr/[lib,include]. Changing the install path with --prefix on the configure step to "%{buildroot}/usr/" will produce some nasty absolute paths in the final build files and rightfully trigger the rpmbuild checker.
- The header files are copied from /usr/local/include/coin-or/ to %{buildroot}/usr/ after doing "make install". This was easier and closer to the upstream build procedure than scrapping through the build tree with a regex for the necessary header files.
- /usr/local/include/coin-or/ is wiped during each build on each machine, so that there are no old header files packed in the RPM.

My biggest takeaway was that it would be very helpful to have dedicated lib and include folders during Ipopt compilation. Installing and/or integrating the program manually (or by any downstream distribution) is way easier if you already have the "build/lib" and "build/include" folders ready to go. Currently the libs are spread over some hidden .libs folders and the headers are part of the src folders.

During "make install" the files have to be gathered anyway, eventually this could already be part of the preceding "make" step. It may already be possible to do this with the more in-depth prefix options from configure, but I think there was no way to distinguish between "install here but link everything as if it will get moved to some other place later".

All in all I'm content with the result ("works for me") and maybe someone else finds this helpful in the future.

[coin-or-Ipopt-3.13.4-0-el8.spec.txt](https://github.com/coin-or/Ipopt/files/6534706/coin-or-Ipopt-3.13.4-0-el8.spec.txt)

**Reply by svigerske** [reply]
I don't really understand why you cannot do a make install and then pack up what has been installed, that is, I don't know what the "nasty absolute paths in the final build files" are. Use `--enable-relocatable` to get nicer pkg-config files.

Installing header files into the `coin-or` subdir is intentional. The pkg-config file helps users to find out what `-I` flag to use to get the Ipopt header files.

**Reply by Zomtir** [reply]
> I don't really understand why you cannot do a make install and then pack up what has been installed

That's exactly what I ended up doing. But this is not 100% foolproof if you build multiple versions on the very same system without properly cleaning up the previous install. Occasionally you'll end up packing some headers/libs/symlinks from the wrong version in the package.

To prevent that from happening, rpmbuild encapsulates the install directory from the compiling machine at a "nasty" location:
```/root/rpmbuild/BUILDROOT/coin-or-Ipopt-{version}-0.el8/usr/```

If you use `..configure --prefix /root/rpmbuild/BUILDROOT/coin-or-Ipopt-{version}-0.el8/usr/` this path will also be part of some binaries, the libipopt.la and more. The rpmbuild checker complains if it can "grep" any BUILDROOT string in the built package.

enable-relocatable might be sufficient to circumvent that, but I overlooked the parameter unfortunately.

**Reply by svigerske** [reply]
I hope that the [`DESTDIR` mechanism](https://www.gnu.org/software/automake/manual/html_node/DESTDIR.html) still works.
With that, you can let the build system think that it installs into `/usr`, but then actually install somewhere else. So you would do
```
../configure --prefix=/usr
make
make DESTDIR=/root/rpmbuild/BUILDROOT/coin-or-Ipopt-{version}-0.el8/ install
```
